include/deal.II/base/vectorization.h

```diff
@@ -4720,6 +4720,27 @@ inline DEAL_II_ALWAYS_INLINE VectorizedArray<Number, width>
   return tmp *= v;
 }
 
+/**
+ * Multiplication of two vectorized arrays with operator *.
+ * This variant permits the input arrays to have different underlying types.
+ *
+ * @relatesalso VectorizedArray
+ */
+template <typename Number1,
+          typename Number2,
+          std::size_t width,
+          typename = typename std::enable_if<
+            !std::is_same<Number1, Number2>::value>::type>
+inline DEAL_II_ALWAYS_INLINE
+  VectorizedArray<typename ProductType<Number1, Number2>::type, width>
+  operator*(const VectorizedArray<Number1, width> &u,
+            const VectorizedArray<Number2, width> &v)
+{
+  using Result_t =
+    VectorizedArray<typename ProductType<Number1, Number2>::type, width>;
+  return Result_t(u) * Result_t(v);
```

Comment:
Hi @kronbichler and @peterrum. Could you give me some advice here? I know that there's no constructor that converts a `VectorizedArray` of one type to another. Is it possible to create such a constructor in an optimal way, or would a utility function like this
```c++
template<typename ScalarTypeOut, typename ScalarTypeIn, std::size_t width>
auto convert(const VectorizedArray<ScalarTypeIn, width> &in)
{
  VectorizedArray<ScalarTypeOut, width> out;
  DEAL_II_OPENMP_SIMD_PRAGMA
  for (unsigned int v = 0; v < width; v++)
    out[v] = in[v];
  return out;
}
```
be the best option here?

Reply:
From a technical point of view, the available fast conversion schemes are `double` -> `float` and `float` -> `double`. I think doing what you suggest makes sense, and good compilers might even do the optimal single-instruction conversion. For all other types, the conversion you suggest is the appropriate choice anyway. So I think this looks good.

Reply:
Thanks for your feedback, @kronbichler. I've added a free function to do the conversion, because I realised that we can avoid a copy when the input scalar type is the same as the output scalar type. About the fast conversions: there are already some specialisation of the math operators that you implemented for some mixed types. I left those as is, and we could look at adding some more in the future if you feel it necessary?